tests/unit/test_metadata.py

```diff
@@ -124,3 +125,109 @@ def test_get(
     data = metadata.get(path=tmp_path)
 
     assert data == expected_metadata
+
+
+@pytest.mark.parametrize(
+    "charmcraft_yaml_content, expected_error_msg_contents",
+    [
+        pytest.param("", ("empty", metadata.CHARMCRAFT_FILENAME), id="malformed"),
+        pytest.param(
+            "malformed: yaml:", ("malformed", metadata.CHARMCRAFT_FILENAME), id="malformed"
+        ),
+        pytest.param("value 1", ("not", "mapping", metadata.CHARMCRAFT_FILENAME), id="not dict"),
+    ],
+)
+def test_get_charmcraft_yaml_malformed(
+    charmcraft_yaml_content: str, expected_error_msg_contents: tuple[str, ...], tmp_path: Path
+):
+    """
+    arrange: given directory with charmcraft.yaml that is malformed
+    act: when get is called with the directory
+    assert: then InputError is raised.
+    """
+    create_charmcraft_yaml(content=charmcraft_yaml_content, path=tmp_path)
+
+    with pytest.raises(exceptions.InputError) as exc_info:
+        metadata.get(path=tmp_path)
+
+    assert_substrings_in_string(expected_error_msg_contents, str(exc_info.value).lower())
+
+
+@pytest.mark.parametrize(
+    "charmcraft_yaml_content, expected_error_msg_contents",
+    [
+        pytest.param("key: value", ("could not find", metadata.CHARMCRAFT_NAME_KEY)),
+        pytest.param(
+            f"{metadata.CHARMCRAFT_NAME_KEY}: 1",
+            (
+                "invalid value for",
+                metadata.CHARMCRAFT_NAME_KEY,
+            ),
+        ),
+        pytest.param(
+            f"{metadata.CHARMCRAFT_NAME_KEY}: name\n{metadata.CHARMCRAFT_LINKS_KEY}: 1",
+            ("invalid value for", metadata.CHARMCRAFT_LINKS_KEY),
+        ),
+        pytest.param(
+            (
+                f"{metadata.CHARMCRAFT_NAME_KEY}: name\n"
+                f"{metadata.CHARMCRAFT_LINKS_KEY}:\n {metadata.CHARMCRAFT_LINKS_DOCS_KEY}: 1"
+            ),
+            ("invalid value for", metadata.CHARMCRAFT_LINKS_DOCS_KEY),
+        ),
+    ],
+)
+def test_get_invalid_charmcraft(
+    charmcraft_yaml_content: str, expected_error_msg_contents: tuple[str, ...], tmp_path: Path
+):
+    """
+    arrange: given charmcraft with missing required keys or invalid value
+    act: when get is called with the directory
+    assert: InputError is raised with missing keys info.
+    """
+    create_charmcraft_yaml(content=charmcraft_yaml_content, path=tmp_path)
+
+    with pytest.raises(exceptions.InputError) as exc_info:
+        metadata.get(path=tmp_path)
+
+    assert_substrings_in_string(expected_error_msg_contents, str(exc_info.value).lower())
+
+
+# pylint currently does not understand walrus operator perfectly yet
+# pylint: disable=unused-variable,undefined-variable
```

Comment:
I know this is not easy to know, there is actually an alternative pattern for how to do this which has partially been implemented and we have been migrating tests to that pattern over time. See here for an example: https://github.com/canonical/discourse-gatekeeper/blob/5c6f9bedadf2137b9e6f93ed2b50fcdb58fca084/tests/unit/test_check.py#L17

Reply:
It would be great if you could update the all the tests at least in this file to use the new pattern, no need to do them all, just keep making progress 🙂

Reply:
sure, I will change that file. I have already seen that way, but I did not know that was the recommended way. TBH I found the walrus operator a bit smart but nice!

Reply:
Done!